spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteReachableFilesSparkAction.java

```diff
@@ -132,7 +135,20 @@ private Dataset<FileInfo> reachableFileDS(TableMetadata metadata) {
   }
 
   private DeleteReachableFiles.Result deleteFiles(Iterator<FileInfo> files) {
-    DeleteSummary summary = deleteFiles(deleteExecutorService, deleteFunc, files);
+    DeleteSummary summary;
+    if (deleteFunc != null || !(io instanceof SupportsBulkOperations)) {
+      if (deleteFunc == null) {
+        LOG.info("Table IO does not support Bulk Operations. Using non-bulk deletes.");
+        deleteFunc = defaultDelete;
```

Comment:
Hm, it seems a bit odd to init this variable here. Also, setting this will change the behavior if the action is called multiple times. Can we pass delete func explicitly?

```
if (deleteFunc == null && io instanceof SupportsBulkOperations) {
  LOG.info("Table IO supports bulk operations, using bulk deletes");
  summary = deleteFiles((SupportsBulkOperations) io, files);
} else {
  if (deleteFunc == null) {
    LOG.info("Table IO does not support bulk operations, using non-bulk deletes");
    summary = deleteFiles(deleteExecutorService, io::deleteFile, files);
  } else {
    LOG.info("Custom delete function provided, using non-bulk deletes");
    summary = deleteFiles(deleteExecutorService, deleteFunc, files);
  }
}
```

Reply:
If we follow the snippet above, we can remove `defaultDelete` completely.